mitxgraders/formulagrader/matrixgrader.py

```diff
@@ -59,3 +85,78 @@ def check_response(self, answer, student_input, **kwargs):
             else:
                 return {'ok': False, 'msg': err.message, 'grade_decimal': 0}
         return result
+
+    @staticmethod
+    def validate_student_input_shape(student_input, expected_shape, detail):
+        """
+        Checks that student_input has expected_shape and raises a ShapeError
+        if it does not.
+
+        Arguments:
+            student_input (number | MathArray): The numerically-sampled student
+                input
+            expected_shape (tuple): A numpy shape tuple
+            detail (None|'shape'|'type') detail-level of ShapeError message
+        """
+        try:
+            input_shape = student_input.shape
+        except AttributeError:
+            if isinstance(student_input, Number):
+                input_shape = tuple()
+            else:
+                raise
+
+        if expected_shape == input_shape:
+            return True
+
+        if detail is None:
+            raise ShapeError('')
+
+        if detail == 'shape':
+            expected = MathArray.get_description(expected_shape)
+            received = MathArray.get_description(input_shape)
+        else:
+            expected = MathArray.get_shape_name(len(expected_shape))
+            received = MathArray.get_shape_name(len(input_shape))
+
+        if detail != 'shape' and expected == received:
+            msg = ("Expected answer to be a {0}, but input is a {1} "
+                   "of incorrect shape".format(expected, received))
+        else:
+            msg = ("Expected answer to be a {0}, but input is a {1}"
+                .format(expected, received))
+
+        raise ShapeError(msg)
+
+    Utils = namedtuple('Utils', ['tolerance', 'within_tolerance', 'validate_shape'])
+
+    def get_comparer_utils(self):
+        """Get the utils for comparer function."""
+        def _within_tolerance(x, y):
+            return within_tolerance(x, y, self.config['tolerance'])
+        def _validate_shape(student_input, shape):
+            detail = self.config['answer_shape_mismatch']['msg_detail']
+            return self.validate_student_input_shape(student_input, shape, detail)
+
+        return self.Utils(tolerance=self.config['tolerance'],
+                          within_tolerance=_within_tolerance,
+                          validate_shape=_validate_shape)
+
+    def default_equality_comparer(self, comparer_params, student_input, utils):
+        """
+        Default comparer function.
+
+        Assumes comparer_params is just the single expected answer wrapped in a list.
+        """
+        expected_input =comparer_params[0]
+        try:
+            # in numpy, scalars have empty tuples as their shapes
+            shape = tuple() if isinstance(expected_input, Number) else expected_input.shape
+            utils.validate_shape(student_input, shape)
+        except ShapeError as err:
```

Comment:
Can we move this try/except block into MatrixGrader, so that comparers can just raise a ShapeError and it's treated appropriately? Open to suggestions here.